### ISSUE TITLE
Losses to implement and losses naming conventions

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn_common#38]
Some modules we should implement:

* `CrossEntropy` or `CE`. Should this cover both dense and sparse targets, or do we want a separate module for the sparse case, like `SparseCrossEntropy` or so? Should this potentially also allow for logits? log-probs?
* `KL` or `KullbackLeiblerDivergence`
* `BinaryCrossEntropy` or `BCE`
* `L2Dist` (absolute or mean?) Or `MSE` or `MeanSquarredError`? (The mean reduction is over the feature axis. Not over time or batch.)
* `L1Dist` (absolute or mean?) Or `MeanL1Dist`?
* `Ctc` or `CTC` or `CtcLogProb`
* `CosineSimilarity`

I don't like the naming of the [PyTorch losses](https://pytorch.org/docs/stable/nn.html) too much here.
They have the postfix `Loss` on all of them, although these modules are generic and not necessarily just for loss computation (although that's probably their most common usage).
Also `CrossEntropyLoss` is actually log-softmax + CE together. So very much like the TF `tf.nn.sparse_softmax_cross_entropy_with_logits`.
And there is a separate `NLLLoss`. Which is just like `CrossEntropyLoss` but it doesn't take logits but log-prob instead. I find this naming to be confusing.

Also the question is how we should handle things like label smoothing. On RETURNN side (and also in TF), it is just an option to the CE-loss. On PyTorch side, it is not implemented yet as part of the official PyTorch API. Some background [here](https://stackoverflow.com/questions/55681502/label-smoothing-in-pytorch). It was only very recently added (https://github.com/pytorch/pytorch/issues/7455, https://github.com/pytorch/pytorch/pull/63122). This also adds it as an option `label_smoothing` to `CrossEntropyLoss`. An alternative would be that the user makes this more explicit, like:
```
target_prob_smooth = smooth_one_hot(targets, label_prob=0.9)
loss = cross_entropy(target_prob_smooth, out_prob)
```
Although label smoothing has become very common, so maybe it makes sense to have this also just as an option.

Note also that the loss accumulation over the dataset and handling of calculating the correct average (mean) is handled by RETURNN. All such losses would just yield a vector of shape [B] or [B,T].

**Comment by Atticus1806** [comment]
>Do we want cross_entropy to work on different input types (logits, log prob, prob)? Or have separate functions for each case? Like cross_entropy_logits etc. Not sure...

We want `rc` to be straight forward and have as little hidden stuff as possible. This is why I think having explicit names like `cross_entropy_logits` would be the most straight forward for the user to use. If we want to restrict ourselves to one function though, I would vote for forcing the user to explicitly set the "Flag" (instead of it having a default value, which again would hide things in my opinion).

>How would we handle label smoothing?
In general this comes down to the argument I think how much of the code is easy for the user to think of and where the border for this is. I think we should answer these kind of questions always by asking ourselves: "Is it reasonable to assume that someone who will use this can write it him/herself or would it require quite some knowledge/searching to come up with?" Once we agree on one of the two this then directly leads to whether we should implement it I think.

>... the problem of adding more and more options to the base cross_entropy function, which is sth we should try to avoid. 

Agreed, so either 2nd option or 3rd in my opinion. For label smoothing I might tend to 2 right now, but maybe others can also comment.

>What functions do we want to introduce, and using what names?

If we name functions error doesn't this go into the same direction as giving functions the `_loss` suffix? Since `error` in the end like `loss` is an interpretation of the difference?

**Comment by albertz** [comment]
> `cross_entropy_logits` ...

Ok. So we would have 3 functions then:
* `cross_entropy_logits`
* `cross_entropy_log_prob`
* `cross_entropy_prob`

> how much of the code is easy for the user ... someone ...

This can vary a lot, depending on the user and his/her background. Some people also might not even know about cross entropy and they just want to train it in whatever way.

In practice, people will probably anyway still use some small code snippets from someone. This can be in the documentation, some small code snippet how you would do label smoothing (just the code snippet I posted above).

Automatic code completion and IDE support is another thing though, where your IDE can easily show you the variants. And as label smoothing is really so common now, maybe we can anyway also add `cross_entropy_logits_label_smoothing` (but not for the other CE function variants, such that we don't get the combinatoric explosion)

Or I wonder how good GitHub Copilot will be to come up with the snippet when I just put the prefix `# cross entropy with label smoothing`...

**Comment by albertz** [comment]
Just as a random reference, Fairseq has [`LabelSmoothedCrossEntropyCriterion`](https://github.com/pytorch/fairseq/blob/fcca32258c8e8bcc9f9890bf4714fa2f96b6b3e1/fairseq/criterions/label_smoothed_cross_entropy.py#L56).

**Comment by albertz** [comment]
Another question: Should it be `cross_entropy_logits(target_prob, out_logits)` or `cross_entropy_logits(out_logits, target_prob)`?

When you look at the [mathematical definition of cross entropy `H(p,q)`](https://en.wikipedia.org/wiki/Cross_entropy), the `p` is the target distribution and `q` is the output.
This is the reverse of how it is defined in [PyTorch `CrossEntropyLoss`](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) which gets `(input, target)`.
However it matches TensorFlow `sparse_softmax_cross_entropy_with_logits` which gets `(labels, logits)`.

**Comment by albertz** [comment]
Note: We also should add (non-differential) functions for frame-error, edit-distance, etc. (See #57.)

**Comment by albertz** [comment]
Related is also #17 on dim tags. Many of these losses will reduce over some axis, usually the feature dim axis. But maybe that should be explicit?

**Comment by albertz** [comment]
> Ok. So we would have 3 functions then:
> 
> * `cross_entropy_logits`
> * `cross_entropy_log_prob`
> * `cross_entropy_prob`

I want to ask again about whether these functions should deal with both dense and sparse targets, or whether this should be separated and explicit as well? If this is separate, then we would have now already 6 functions. Is this good? And maybe there are other things?

Also note, if we don't separate sparse inputs, then we need to implement it such that it can handle both dense and sparse inputs. This implies that we need one of these:

- #47
- https://github.com/rwth-i6/returnn/issues/741
- Or an explicit option `dense_target=False` or so.

**Comment by albertz** [comment]
Note that `DotLayer` supports sparse inputs now (returnn#741).

And we also have shape information and `sparse_dim` available (#47).

But I would just use `DotLayer` for simplicity.

However, we should somehow make sure that it uses `sparse_softmax_cross_entropy_with_logits` for the standard case for efficiency reasons. And also make sure it is only calculated on the non-padded frames (#68), again for efficiency.

**Comment by albertz** [comment]
We have `cross_entropy` now.

I decided to go with a (required) option `...type` which is "probs", "log-probs" or "logits".

It also wraps `sparse_softmax_cross_entropy_with_logits` when possible.
Otherwise it uses `dot` (both for dense or sparse).

It also has a `label_smoothing` option but not sure if that was a good idea. As mentioned before, it should stay very minimal. But on the other side, label smoothing is very common. But maybe I remove that again.

One important aspect which is missing is the flattening (#68). This is also for efficiency because we don't want to do unnecessary softmax computations. But in any case, we should deal with this separately (#68).

**Comment by albertz** [comment]
I think we now have a good starting point with `cross_entropy`, `binary_cross_entropy`, `mean_squared_difference`, `mean_absolute_difference`.

One (for us) relevant remaining loss is CTC. Here is a bit the question on how generic we want to have it. In any case there should be a `ctc_loss` corresponding to the original CTC loss, wrapping `returnn.tf.native_op.ctc_loss` (same as the TF CTC loss but faster). But then we also might want to have a more generic full sum loss, for any target FSA (e.g. generated on-the-fly by RASR). And the `ctc_loss` can maybe directly be based on that (internally in RETURNN it's also like that anyway).

Once we have `ctc_loss`, I think this issue can be closed and any further loss can be implemented later once it is needed.

**Comment by albertz** [comment]
We also have `ctc_loss` now, directly wrapping the newly introduced `CtcLossLayer`.

So I will close this now. In case sth is buggy, please open a new issue. In case you want to have a specific missing loss function, please open a new issue.
